### PR TITLE
Added a metadata lineage processor and a MetadataLineageEvent.avsc

### PR DESCRIPTION
**Pull request by shpark76**

### CODE REVIEW

File: wherehows-dao/src/main/java/wherehows/dao/table/LineageDao.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+public class LineageDao {
+
+  /**
+   * Create lineage dataset that requested the lineage via Kafka lineage event.
+   * @param datasetLineages List of lineages
+   * @return return process result as true/false
+   */
+  public Boolean createLineages(List<DatasetLineage> datasetLineages) {
+    // TODO: write lineage Dao to DB
+    throw new UnsupportedOperationException("Lineage not implemented for open source.");
```
Comment (alyiwang) [inline wherehows-dao/src/main/java/wherehows/dao/table/LineageDao.java]: Can we remove "for open source." with "yet"

File: gradle/scripts/ember.gradle [hunk 2]
```diff
@@ -48,4 +48,4 @@ clean {
   delete 'node_modules'
   delete 'bower_components'
   delete 'dist'
-}
+}
```
Comment (alyiwang) [inline gradle/scripts/ember.gradle]: Could we reset this

File: wherehows-kafka/src/main/java/wherehows/processors/MetadataLineageProcessor.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+  private final LineageDao _lineageDao = DAO_FACTORY.getLineageDao();
+
+  public MetadataLineageProcessor(DaoFactory daoFactory, String producerTopic,
+      KafkaProducer<String, IndexedRecord> producer) {
+    super(daoFactory, producerTopic, producer);
+  }
+
+  /**
+   * Process a MetadataChangeEvent record
```
Comment (alyiwang) [inline wherehows-kafka/src/main/java/wherehows/processors/MetadataLineageProcessor.java]: MetadataLineageEvent

File: wherehows-kafka/src/main/java/wherehows/processors/MetadataLineageProcessor.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+   * @param indexedRecord IndexedRecord
+   * @throws Exception
+   */
+  public void process(IndexedRecord indexedRecord) {
+
+    if (indexedRecord == null || indexedRecord.getClass() != MetadataLineageEvent.class) {
+      throw new IllegalArgumentException("Invalid record");
+    }
+
+    log.info("Processing Metadata Lineage Event record. ");
```
Comment (alyiwang) [inline wherehows-kafka/src/main/java/wherehows/processors/MetadataLineageProcessor.java]: log.debug

File: wherehows-kafka/src/main/java/wherehows/processors/MetadataLineageProcessor.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+    MetadataLineageEvent record = (MetadataLineageEvent) indexedRecord;
+
+    final KafkaAuditHeader auditHeader = record.auditHeader;
+    if (auditHeader == null) {
+      log.warn("MLE: MetadataLineageEvent without auditHeader, abort process. " + record.toString());
+      return;
+    }
+    log.debug("MLE: string : " + record.toString());
```
Comment (alyiwang) [inline wherehows-kafka/src/main/java/wherehows/processors/MetadataLineageProcessor.java]: Can we combine some of those logs, it seems a lot of logging here.

File: wherehows-kafka/src/main/java/wherehows/processors/MetadataLineageProcessor.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    log.info("MLE: TS: " + auditHeader.time);
+    log.debug("MLE: agentName: " + record.type.name());
+    log.debug("MLE: deploymentDetail: " + record.deploymentDetail.toString());
+    log.debug("MLE: jobExecution: " + record.jobExecution.toString());
+    log.info("MLE: lineage: " + record.lineage.toString());
+
+    if (record.lineage == null || record.lineage.size() <= 0) {
```
Comment (alyiwang) [inline wherehows-kafka/src/main/java/wherehows/processors/MetadataLineageProcessor.java]: maybe just == 0, no need <= 0

File: wherehows-kafka/src/main/java/wherehows/processors/MetadataLineageProcessor.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+      throw new IllegalArgumentException("Invalid lineage record");
```
Comment (alyiwang) [inline wherehows-kafka/src/main/java/wherehows/processors/MetadataLineageProcessor.java]: Can be more specific: "No Lineage info in record"

File: wherehows-kafka/src/main/java/wherehows/processors/MetadataLineageProcessor.java (hunk 5 above)
Comment (shpark76) [reply wherehows-kafka/src/main/java/wherehows/processors/MetadataLineageProcessor.java]: done

File: gradle/scripts/ember.gradle (hunk 2 above)
Comment (shpark76) [reply gradle/scripts/ember.gradle]: it's removed unnecessary white space.

File: wherehows-kafka/src/main/java/wherehows/processors/MetadataLineageProcessor.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+  private final LineageDao _lineageDao = DAO_FACTORY.getLineageDao();
+
+  public MetadataLineageProcessor(DaoFactory daoFactory, String producerTopic,
+      KafkaProducer<String, IndexedRecord> producer) {
+    super(daoFactory, producerTopic, producer);
+  }
+
+  /**
+   * Process a MetadataKineageEvent record
```
Comment (alyiwang) [inline wherehows-kafka/src/main/java/wherehows/processors/MetadataLineageProcessor.java]: typo: Lineage

Comment (shpark76) [reply wherehows-kafka/src/main/java/wherehows/processors/MetadataLineageProcessor.java]: Thanks for catching up.

**Pull request merged**